### PR TITLE
Add health checker for cluster management module

### LINKED ISSUE #5633: Add health check for Storage/Cluster management module
In the previous releases, @hanahmily added the health check module and its implementation for the storage modules(H2, MySQL, and ES), we need to add a similar thing for InfluxDB.

Also, for cluster management, we need to add this too. If 
1. the registration fails 
1. can't get the instance list 
1. can't get itself, then
1. get multiple itself(such as multiple 127.0.0.1/0.0.0.0) in the instance list
The cluster management should be listed as an error 

The health check could be verified through CLI command, https://github.com/apache/skywalking-cli#checkhealth

@xbkaishui Are you going to do this?

### PR DESCRIPTION
**Pull request by xbkaishui**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.
resolved #5633

### CODE REVIEW

**Review by wu-sheng — CHANGES_REQUESTED** [review]
From my understanding, `DelegatedHealthChecker` is designed for classes without accessibility to the telemetry API, such as ElasticSearch client is hosted in the `library-client`.

In the cluster case, the telemetry API(`HealthCheckMetrics`) is better to use directly.

**Comment by xbkaishui** [comment]
ok will change

**Comment by xbkaishui** [comment]
done

File: oap-server/server-cluster-plugin/pom.xml [hunk 1]
```diff
@@ -47,5 +47,10 @@
             <artifactId>library-util</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.skywalking</groupId>
+            <artifactId>library-client</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (wu-sheng) [inline oap-server/server-cluster-plugin/pom.xml:54]: You should remove this dependency totally.

Comment (xbkaishui) [reply oap-server/server-cluster-plugin/pom.xml]: done

**Review by wu-sheng — COMMENTED** [review]
Please remove `import org.apache.skywalking.oap.server.library.util.HealthChecker`. As in the module provider, we don't need that.

**Comment by xbkaishui** [comment]
Hi @wu-sheng ,  seems the failed e2e test are caused by unhealthy status, I can't find any error log about this, it seems the health check is not stable,  how to fix this ? 
```
14:03:08.217 INFO  🐳 [docker-compose].processLine (35) - ERROR: for oap2  Container "dcc4c103dba6" is unhealthy.
14:03:08.217 INFO  🐳 [docker-compose].processLine (35) - ERROR: for provider1  Container "dcc4c103dba6" is unhealthy.
```

**Comment by wu-sheng** [comment]
>  seems the failed e2e test are caused by unhealthy status, I can't find any error log about this, it seems the health check is not stable, how to fix this ?

I have rerun, maybe just unstable test.

File: oap-server/server-cluster-plugin/cluster-consul-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/consul/ConsulCoordinator.java [hunk 2]
```diff
@@ -27,19 +27,25 @@
...
 import org.apache.skywalking.oap.server.core.cluster.RemoteInstance;
+import org.apache.skywalking.oap.server.core.cluster.ServiceQueryException;
 import org.apache.skywalking.oap.server.core.cluster.ServiceRegisterException;
 import org.apache.skywalking.oap.server.core.remote.client.Address;
+import org.apache.skywalking.oap.server.library.client.healthcheck.HealthCheckable;
 import org.apache.skywalking.oap.server.library.util.CollectionUtils;
+import org.apache.skywalking.oap.server.library.util.HealthChecker;
 
-public class ConsulCoordinator implements ClusterRegister, ClusterNodesQuery {
+public class ConsulCoordinator implements ClusterRegister, ClusterNodesQuery, HealthCheckable {
```
Comment (wu-sheng) [inline oap-server/server-cluster-plugin/cluster-consul-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/consul/ConsulCoordinator.java:42]: You are still using the ` org.apache.skywalking.oap.server.library.util.HealthChecker` and `org.apache.skywalking.oap.server.library.client.healthcheck.HealthCheckable`.

Comment (xbkaishui) [reply oap-server/server-cluster-plugin/cluster-consul-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/consul/ConsulCoordinator.java]: done

File: oap-server/server-cluster-plugin/cluster-consul-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/consul/ConsulCoordinator.java [hunk 3]
```diff
@@ -49,22 +56,32 @@ public ConsulCoordinator(ClusterModuleConsulConfig config, Consul client) {
...
+            // Discover only "passing" nodes
+            List<ServiceHealth> nodes = healthClient.getHealthyServiceInstances(serviceName).getResponse();
+            if (CollectionUtils.isNotEmpty(nodes)) {
+                nodes.forEach(node -> {
+                    if (!Strings.isNullOrEmpty(node.getService().getAddress())) {
+                        Address address = new Address(node.getService().getAddress(), node.getService().getPort(), false);
+                        if (address.equals(selfAddress)) {
+                            address.setSelf(true);
+                        }
+                        remoteInstances.add(new RemoteInstance(address));
```
Comment (wu-sheng) [inline oap-server/server-cluster-plugin/cluster-consul-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/consul/ConsulCoordinator.java:71]: If you found `127.0.0.1` or `localhost` exists in the node list(list#size > 1), then the node should be labeled as unhealthy.
FYI @hanahmily @kezhenxu94 Make sense?

Comment (wu-sheng) [reply oap-server/server-cluster-plugin/cluster-consul-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/consul/ConsulCoordinator.java:71]: Same thing should be checked in every coordinator implementation. I noted that the users usually set up the cluster mode incorrectly, then, blame SkyWalking insert duplicated data or statistic not accurate.

Comment (xbkaishui) [reply oap-server/server-cluster-plugin/cluster-consul-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/consul/ConsulCoordinator.java:71]: got it

Comment (xbkaishui) [reply oap-server/server-cluster-plugin/cluster-consul-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/consul/ConsulCoordinator.java:73]: Done, current implementation will ignore coordinator  standalone mode

File: oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/HealthCheckUtil.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ */
+
+package org.apache.skywalking.oap.server.library.util;
+
+import com.google.common.collect.Sets;
+
+import java.util.Set;
+
+public class HealthCheckUtil {
```
Comment (wu-sheng) [inline oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/HealthCheckUtil.java:25]: Please move this class to the same folder of `ClusterModule`, as where it belongs. Then rename it to `OAPNodeChecker`.

File: oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/HealthCheckUtil.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    private static final Set<String> UN_HEALTH_ADDRESS = Sets.newHashSet("127.0.0.1", "localhost");
```
Comment (wu-sheng) [inline oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/HealthCheckUtil.java:26]:
```suggestion
    private static final Set<String> ILLEGAL_NODE_ADDRESS_IN_CLUSTER_MODE = Sets.newHashSet("127.0.0.1", "localhost");
```

Comment (xbkaishui) [reply oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/HealthCheckUtil.java]: done

File: oap-server/server-cluster-plugin/cluster-etcd-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/etcd/EtcdCoordinator.java [hunk 6]
```diff
@@ -51,6 +59,9 @@
 
     private static final Integer KEY_TTL = 45;
 
+    @Setter
+    private HealthCheckMetrics healthChecker;
```
Comment (wu-sheng) [inline oap-server/server-cluster-plugin/cluster-etcd-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/etcd/EtcdCoordinator.java:63]: As this is required in the process, please don't use the `setter`. Initial this in the constructor.

File: oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeCheckerTest.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+import static org.hamcrest.CoreMatchers.is;
+
+public class OAPNodeCheckerTest {
+
+    @Test
+    public void hasUnHealthAddressFalse() {
+        Set<String> address = Sets.newHashSet("123.23.4.2");
+        boolean flag = OAPNodeChecker.hasUnHealthAddress(address);
+        Assert.assertThat(flag, is(false));
```
Comment (wu-sheng) [inline oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeCheckerTest.java:35]: There is `Assert.assertFalse()`

File: oap-server/server-cluster-plugin/cluster-etcd-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/etcd/EtcdCoordinator.java [hunk 8]
```diff
@@ -73,15 +83,30 @@ public EtcdCoordinator(ClusterModuleEtcdConfig config, EtcdClient client) {
...
+            if (remoteInstances.size() > 1) {
+                Set<String> remoteAddressSet = remoteInstances.stream().map(remoteInstance ->
+                        remoteInstance.getAddress().getHost()).collect(Collectors.toSet());
+                boolean hasUnHealthAddress = OAPNodeChecker.hasUnHealthAddress(remoteAddressSet);
+                if (hasUnHealthAddress) {
+                    this.healthChecker.unHealth(new ServiceQueryException("found 127.0.0.1 or localhost in cluster mode"));
+                } else {
+                    List<RemoteInstance> selfInstances = remoteInstances.stream().
+                            filter(remoteInstance -> remoteInstance.getAddress().isSelf()).collect(Collectors.toList());
+                    if (CollectionUtils.isNotEmpty(selfInstances) && selfInstances.size() == 1) {
```
Comment (wu-sheng) [inline oap-server/server-cluster-plugin/cluster-etcd-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/etcd/EtcdCoordinator.java:98]: Why `&& selfInstances.size() == 1`? Are you planning to check duplicate addresses? If so, please move this login into `OAPNodeChecker`, all coordinator implementations should check this too, right?

Comment (xbkaishui) [reply oap-server/server-cluster-plugin/cluster-etcd-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/etcd/EtcdCoordinator.java:98]: Yes, changed

File: oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeCheckerTest.java (hunk 7 above)
Comment (xbkaishui) [reply oap-server/server-core/src/test/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeCheckerTest.java:35]: done

File: oap-server/server-cluster-plugin/cluster-etcd-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/etcd/EtcdCoordinator.java (hunk 6 above)
Comment (xbkaishui) [reply oap-server/server-cluster-plugin/cluster-etcd-plugin/src/main/java/org/apache/skywalking/oap/server/cluster/plugin/etcd/EtcdCoordinator.java:63]: done

**Review by wu-sheng — COMMENTED** [review]
Recommend to read `RemoteClientManager#L123`, the duplicated node will be distinct. So, the duplicate instance nodes would not happen in the core.
NOTE, for many coordinators like ZK, if the OAP reboot, there is a short of time, the register information would be duplicated, as the old temp node has not been removed/expired.

**Comment by xbkaishui** [comment]
> Recommend to read `RemoteClientManager#L123`, the duplicated node will be distinct. So, the duplicate instance nodes would not happen in the core.

I am not checking the duplicated node, maybe the method name is not clear , I just checking belows condtions: 
get multiple itself(such as multiple 127.0.0.1/0.0.0.0) in the instance list

**Comment by wu-sheng** [comment]
> get multiple itself(such as multiple 127.0.0.1/0.0.0.0) in the instance list

You have two checking methods, this is just one thing.

**Comment by xbkaishui** [comment]
Sorry, can you explain more, I am not sure what's is wrong

**Comment by xbkaishui** [comment]
> Recommend to read `RemoteClientManager#L123`, the duplicated node will be distinct. So, the duplicate instance nodes would not happen in the core.
> NOTE, for many coordinators like ZK, if the OAP reboot, there is a short of time, the register information would be duplicated, as the old temp node has not been removed/expired.

Done, please help to review again, thanks

**Comment by wu-sheng** [comment]
We begin the 8.2.0 release process. The PR will be held a little longer.

**Comment by xbkaishui** [comment]
ok, got it

**Comment by xbkaishui** [comment]
Hi @wu-sheng , can you help to review again?

**Comment by wu-sheng** [comment]
I am on vacation, will be back on this at Sunday.

**Comment by xbkaishui** [comment]
got it

**Review by wu-sheng — APPROVED** [review]
LGTM. @xbkaishui Could you prepare another pull request to enhance the setup document? Talking about using CLI health check command?

**Review by wu-sheng — CHANGES_REQUESTED** [review]
@xbkaishui @kezhenxu94 Please recheck on the e2e, seems this change breaks the tests. I have re-run several times.

**Comment by xbkaishui** [comment]
Ok, the error log is not so clear, how can I run the e2e test in my local env?   may need your help

**Comment by xbkaishui** [comment]
> setup

OK, will do this later, thanks

**Comment by kezhenxu94** [comment]
@xbkaishui this may be the cause

image

**Comment by xbkaishui** [comment]
Thanks, it seems in prepare stage, I can't call getManager().find method, will fix this

**Comment by xbkaishui** [comment]
I found the dependency circle problem, the core module remote manager dependency on cluster module, when called start method

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeChecker.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * Check the remote instance healthiness, set health to false for bellow conditions:
+     * 1.can't get the instance list
+     * 2.can't get itself
+     * 3.check for illegal node in cluster mode such as 127.0.0.1, localhost
+     *
+     * @param remoteInstances all the remote instances from cluster
+     * @return true health false unHealth
+     */
+    public static boolean isHealth(List<RemoteInstance> remoteInstances) {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeChecker.java:49]: Make the return value as an object. then you could get a boolean with the reason.
The reason could include which illegal address is found.

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeChecker.java:49]: Then add a new method in `HealthCheckMetrics`, `unHealth(String message)`(note, the message is just for log like exception). Creating exception cause unnecessary resource, and the `HealthQueryService` and `HealthCheckerProvider#notifyAfterCompleted` don't care the exception at all.

**Review by wu-sheng — COMMENTED** [review]
Generally good, need a little more polish.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeChecker.java (hunk 9 above)
Comment (xbkaishui) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeChecker.java:49]: Ok, sounds better to me

Comment (xbkaishui) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/cluster/OAPNodeChecker.java]: done, please help to check

**Comment by wu-sheng** [comment]
You have conflicting, please recheck.

**Comment by xbkaishui** [comment]
rebase has some issues, I have submit a new pr https://github.com/apache/skywalking/pull/5784 
will close this

**Pull request closed without merging**